crates/dodgy_3d/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+        let tangent_ring_triangle_leg_squared =
+          distance_squared - sum_radius_squared;
+
+        let squared_distance_between_rays = relative_neighbour_position
+          .cross(relative_agent_velocity)
+          .length_squared();
+
+        // Use the Pythagorean theorem to solve for the time when the
+        // relative_agent_velocity ray has the "correct" distance from the
+        // origin for where the intersection point should be. The equation is
+        // roughly:
+        // (relative_agent_velocity + t * relative_neighbour_position) ^ 2
+        //   = squared_distance_between_rays / tangent_ring_triangle_leg_squared
+
+        let a = relative_neighbour_position.length_squared();
+        // Note: we *should* multiply this by 2, but we can actually just factor
+        // the two out to skip a few multiplications.
+        let b = relative_neighbour_position.dot(relative_agent_velocity);
+        let c = relative_agent_velocity.length_squared()
+          - squared_distance_between_rays / tangent_ring_triangle_leg_squared;
+        // Always choose the negative solution, since we know the intersection
+        // point must be behind us (since if it was ahead, we should have
+        // projected to the cutoff sphere instead).
+        let t = (-b - (b * b - a * c).sqrt()) / a;
+
+        vo_normal = (relative_agent_velocity + t * relative_neighbour_position)
+          .normalize_or_zero();
+        let distance_to_plane = Plane { normal: vo_normal, point: Vec3::ZERO }
+          .signed_distance_to_plane(relative_agent_velocity);
+        inside_vo = distance_to_plane < 0.0;
+        relative_velocity_projected_to_vo =
+          relative_agent_velocity - distance_to_plane * vo_normal;
+      }
+    } else {
+      // Collision. Project on cut-off sphere at time `time_step`.
+
+      // Find the velocity such that after `time_step` the agent would be at the
+      // neighbours position.
+      let cutoff_sphere_center = relative_neighbour_position / time_step;
+      let cutoff_sphere_radius = sum_radius / time_step;
+
+      // The direction of the velocity from `cutoff_sphere_center` is therefore
+      // the normal to the velocity obstacle.
+      vo_normal =
+        (relative_agent_velocity - cutoff_sphere_center).normalize_or_zero();
+      // Get the point on the cut-off sphere in that direction (which is the
+      // agent's velocity projected to the sphere).
+      relative_velocity_projected_to_vo =
+        vo_normal * cutoff_sphere_radius + cutoff_sphere_center;
+      inside_vo = true;
+    }
+
+    // As in the paper, `u` is the vector from the relative velocity to the
+    // nearest point outside the velocity obstacle.
+    let u = relative_velocity_projected_to_vo - relative_agent_velocity;
+
+    let responsibility = if inside_vo {
+      self.avoidance_responsibility
+        / (self.avoidance_responsibility + neighbour.avoidance_responsibility)
```

Comment:
Nice. This was not part of the C++ 3D implementation. I imagine it will be useful when using agents as obstacles, which should have zero responsibility.

Reply:
Yup, this was part of the 2D version as well!